cpp/include/rapidsmpf/config.hpp

```diff
@@ -0,0 +1,201 @@
+/**
+ * SPDX-FileCopyrightText: Copyright (c) 2025, NVIDIA CORPORATION & AFFILIATES.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#pragma once
+
+#include <any>
+#include <functional>
+#include <memory>
+#include <mutex>
+#include <stdexcept>
+#include <string>
+#include <unordered_map>
+
+#include <rapidsmpf/error.hpp>
+
+namespace rapidsmpf::config {
+
+/**
+ * @brief Type alias for a factory function that constructs options from strings.
+ *
+ * The factory receives the string representation of an option value and returns
+ * an instance of the option type. If the option is unset, the function receives
+ * an empty string and should either return a meaningful default value or throw
+ * `std::invalid_argument`.
+ *
+ * @note The factory must not access other options, as a lock is held during option
+ * initialization and doing so may cause a deadlock.
+ */
+template <typename T>
+using OptionFactory = std::function<T(std::string const&)>;
+
+namespace detail {
+
+/**
+ * @brief Internal implementation of the `Options` class.
+ *
+ * This class is used internally by `Options` to manage the storage and retrieval
+ * of configuration options. Refer to the `Options` class documentation for details.
+ */
+class OptionsImpl {
+  public:
+    /**
+     * @brief Constructs an `OptionsImpl` instance.
+     *
+     * @param options_as_strings A map of option keys to their string representations.
+     * @param options A map of option keys to their corresponding `Option` objects.
+     */
+    OptionsImpl(
+        std::unordered_map<std::string, std::string> options_as_strings,
+        std::unordered_map<std::string, std::any> options
+    );
+
+    /**
+     * @brief Retrieves a configuration option by key.
+     *
+     * Refer to the `Options::get` method for usage details.
+     *
+     * @tparam T The type of the option to retrieve.
+     * @param key The option key (should be lower case).
+     * @param factory Function to construct the option from a string.
+     * @return Reference to the option value.
+     *
+     * @throws std::invalid_argument If the stored option type does not match T.
+     */
+    template <typename T>
+    T const& get(std::string const& key, OptionFactory<T> factory) {
+        {
+            std::lock_guard<std::mutex> lock(mutex_);
+            if (options_.find(key) == options_.end()) {
+                if (options_as_strings_.find(key) == options_as_strings_.end()) {
+                    options_[key] = std::make_any<T>(factory(""));
+                } else {
+                    options_[key] = std::make_any<T>(factory(options_as_strings_[key]));
+                }
```

Comment:
This can avoid multiple lookups in `options_as_strings_` with something like:


```
auto val = options_as_strings_.find(key)
options_[key] = std::make_any<T>(factory(val == options_as_strings_.end() ? "" : *val));
```
if (auto val = options_as_strings_.find(key)